### PR TITLE
HAProxy: add support for draining connections in maintenace mode

### PR DESCRIPTION
**Pull request by drmrd**
##### SUMMARY
A common workflow when using HAProxy as one's load balancer involves taking down a managed node "gracefully" by (1) placing the server in "drain" mode, wherein it will stop accepting new client connections while serve all existing ones until their natural conclusions, and (2) proceed with server maintenance/redeployment as soon as possible after draining is complete. This pull request seeks to automate this workflow through the `haproxy`. It adds a new `drain` option and setting `state: disabled`, `drain: yes`, and `wait: yes` will together result in the module attempting to (1) set the host's HAProxy status to `DRAIN` and then (2) wait and poll the host (in the usual way specified by `wait_retries` and `wait_interval`) until draining is complete. Such a task succeeds when draining finishes and HAProxy places the host in `MAINT` mode, and it is marked as a failure otherwise.

This implements the feature originally requested in [Issue 521](https://github.com/ansible/ansible-modules-extras/issues/521) of the ansible-modules-extras repo.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
HAProxy Module
(`lib/ansible/modules/net_tools/haproxy.py`)

##### ANSIBLE VERSION
```
ansible 2.3.0.0
  config file =
  configured module search path = Default w/o overrides
  python version = 2.7.13 (default, Dec 17 2016, 23:03:43) [GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)]
```

##### ADDITIONAL INFORMATION
This is a rebased PR of #21420, with commit written by @krislindgren. All implementation credit goes to the original author (along with my many thanks!).

### CODE REVIEW

**Comment by drmrd** [comment]
I took a moment to clean up the original submission's documentation and simplify changes made to the `wait_until_status` function down to one line.

**Comment by gundalow** [comment]
@harlowja @bjolivot That you for reviewing the original PR https://github.com/ansible/ansible/pull/21420 those changes have been moved into this fresh PR, and we'd welcome your review

@krislindgren FYI as the original maintainer of the PR

**Comment by krislindgren** [comment]
This change looks good to me.  I just didn't have any time to rebase the old change after a few different ansible releases and a few months had passed.  Thanks for picking this up and running with it.

**Comment by dmsimard** [comment]
It looks like this has landed through other contributions in https://github.com/ansible/ansible/pull/25115 and https://github.com/ansible/ansible/pull/26396 respectively.

**Comment by krislindgren** [comment]
So whats landed 1.) doesn't work on all versions of haproxy 2.) Will never actually put a server into maintenance mode after x amount of time.  Or once it has 0 sessions.  So you need to implement some osrt of loop on your own (since the module wont let you see the number of active sessions).

@drmrd can you rebase this change on top of what landed to give an actual functioning draining mechanism?

**Comment by drmrd** [comment]
@krislindgren Sure thing! I've been away on vacation but will look into doing so early this week.

**Comment by drmrd** [comment]
I've rebased this PR onto devel. With the changes made by #25115 and #26396, there is now a `drain` state and a `drain` parameter, the latter being only used when the state is set to `disabled`. It might be worth strategizing about the interface if this proves to be confusing, but this does coincide with the two different use cases for changing a server to drain mode with Ansible described in  #25115 and #26396 and by @krislindgren above.

Confusion can probably be avoided by providing enough documentation and examples.

**Comment by drmrd** [comment]
I believe this code is ready for review. It was updated last week to take into account #25115 and #26396, and I've just rebased it to remove the stale_ci label.

**Comment by gundalow** [comment]
@krislindgren Could you please review this? Thanks in advance

**Comment by gundalow** [comment]
@ravibhure Could you please review this, thanks in advance.

File: lib/ansible/modules/net_tools/haproxy.py [hunk 1]
```diff
@@ -315,7 +336,7 @@ def get_state_for(self, pxname, svname):
         r = csv.DictReader(data.splitlines())
         state = tuple(
             map(
-                lambda d: {'status': d['status'], 'weight': d['weight']},
+                lambda d: {'status': d['status'], 'weight': d['weight'], 'scur': d['scur']},
```
Comment (drmrd) [inline lib/ansible/modules/net_tools/haproxy.py:341]: The number of active sessions for a backend server is given by the value of `scur` in [HTTP state checks](https://www.haproxy.com/doc/aloha/7.0/haproxy/healthchecks.html#checking-a-http-service).

**Comment by ravibhure** [comment]
shipit 👍

**Comment by drmrd** [comment]
@ravibhure Thanks for the 👍 ! I just submitted an improved version that ensures HAProxy actually drains backends (as opposed to just disabling them as usual). Fortunately it's only a few additional lines long. Would you mind giving it another look?

File: lib/ansible/modules/net_tools/haproxy.py [hunk 2]
```diff
@@ -385,6 +409,8 @@ def act(self):
         # toggle enable/disbale server
         if self.state == 'enabled':
             self.enabled(self.host, self.backend, self.weight)
+        elif self.state == 'disabled' and self.drain:
```
Comment (drmrd) [inline lib/ansible/modules/net_tools/haproxy.py:412]: I went with this flat style, since it reads better to me here than a nested conditional.

**Comment by drmrd** [comment]
@ravibhure @krislindgren Have you been able to give the most recent version of this PR a look? Not to pester, but I'd like to get it merged in time for 2.4 if possible. Thanks!

**Comment by drmrd** [comment]
@gundalow Is there another dev who would be able to review this PR if @ravibhure and @krislindgren are unavailable? It'd be great to have it merged before tomorrow's code freeze.

**Comment by krislindgren** [comment]
+1, Looks good to me.

**Comment by ravibhure** [comment]
+1

**Comment by gundalow** [comment]
Looks good. Will merge into `devel` and release in `2.4.0` once CI passes

rebuild_merge

**Pull request merged**